File: src/com/intellij/advancedExpressionFolding/expression/Expression.java
package com.intellij.advancedExpressionFolding.expression;

import com.intellij.advancedExpressionFolding.AdvancedExpressionFoldingSettings;
import com.intellij.lang.folding.FoldingDescriptor;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.FoldingGroup;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class Expression {
    public static final FoldingDescriptor[] EMPTY_ARRAY = new FoldingDescriptor[0];

    public static String HIGHLIGHTED_GROUP_POSTFIX = ":highlighting";

    private static final Expression NULL_OBJECT = new Expression() {
        @Override
        public String toString() {
            return "NULL_OBJECT";
        }
    };

    protected @NotNull PsiElement element;
    protected @NotNull TextRange textRange;

    private Expression() {
        super();
    }

    public Expression(@NotNull PsiElement element, @NotNull TextRange textRange) {
        this.element = element;
        this.textRange = textRange;
    }

    @Nullable
    public static Expression ofNullable(@Nullable Expression expression) {
        if (expression == null) {
            if (AdvancedExpressionFoldingSettings.getInstance().getState().getMemoryImprovements()) {
                return NULL_OBJECT;
            }
        }
        return expression;
    }

    @Nullable
    public static Expression getOrNull(@Nullable Expression expression) {
        if (expression == NULL_OBJECT) {
            return null;
        }
        return expression;
    }

    @Override
    public String toString() {
        return element.getText(); // TODO: Use document.getText(textRange)
    }

    public boolean supportsFoldRegions(@NotNull Document document,
                                       @Nullable Expression parent) {
        return false; // TODO no-format: This should be impossible
    }

    public FoldingDescriptor[] buildFoldRegions(@NotNull PsiElement element, @NotNull Document document, @Nullable Expression parent) {
        return EMPTY_ARRAY;
    }

    public FoldingDescriptor[] buildFoldRegions(@NotNull PsiElement element, @NotNull Document document, @Nullable Expression parent,
                                                @Nullable FoldingGroup overflowGroup, @Nullable String overflowLeftPlaceholder, @Nullable String overflowRightPlaceholder) {
        return buildFoldRegions(element, document, parent);
    }

    public boolean isCollapsedByDefault() {
        return true;
    }

    @NotNull
    public TextRange getTextRange() {
        return textRange;
    }

    @NotNull
    public PsiElement getElement() {
        return element;
    }

    public boolean isNested() {
        return !this.textRange.equals(element.getTextRange());
    }

    public boolean isOverflow() {
        return isLeftOverflow() || isRightOverflow();
    }

    public boolean isLeftOverflow() {
        return false;
    }

    public boolean isRightOverflow() {
        return false;
    }

    public boolean isHighlighted() {
        return false;
    }

    public TextRange getHighlightedTextRange() {
        return getTextRange();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Expression that = (Expression) o;

        if (!element.equals(that.element)) return false;
        return textRange.equals(that.textRange);
    }

    @Override
    public int hashCode() {
        int result = element.hashCode();
        result = 31 * result + textRange.hashCode();
        return result;
    }


}

File: src/com/intellij/advancedExpressionFolding/expression/Expression.kt
package com.intellij.advancedExpressionFolding.expression

import com.intellij.lang.folding.FoldingDescriptor
import com.intellij.openapi.editor.Document
import com.intellij.openapi.editor.FoldingGroup
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement

abstract class Expression protected constructor() {

    protected lateinit var elementBacking: PsiElement
        private set
    protected lateinit var textRangeBacking: TextRange
        private set

    protected constructor(element: PsiElement, textRange: TextRange) : this() {
        this.elementBacking = element
        this.textRangeBacking = textRange
    }

    open val element: PsiElement
        get() = elementBacking

    open val textRange: TextRange
        get() = textRangeBacking

    open fun supportsFoldRegions(document: Document, parent: Expression?): Boolean {
        return false
    }

    @Deprecated(
        "Use buildFoldRegionsList instead",
        ReplaceWith("buildFoldRegionsList(element, document, parent, foldings)")
    )
    open fun buildFoldRegions(
        element: PsiElement,
        document: Document,
        parent: Expression?
    ): Array<FoldingDescriptor> {
        return EMPTY_ARRAY
    }

    @Deprecated(
        "Use buildFoldRegionsList instead",
        ReplaceWith(
            "buildFoldRegionsList(element, document, parent, overflowGroup, overflowLeftPlaceholder, overflowRightPlaceholder, foldings)"
        )
    )
    open fun buildFoldRegions(
        element: PsiElement,
        document: Document,
        parent: Expression?,
        overflowGroup: FoldingGroup?,
        overflowLeftPlaceholder: String?,
        overflowRightPlaceholder: String?
    ): Array<FoldingDescriptor> {
        return buildFoldRegions(element, document, parent)
    }

    open fun buildFoldRegionsList(
        element: PsiElement,
        document: Document,
        parent: Expression?,
        foldings: MutableList<FoldingDescriptor>
    ) {
    }

    open fun buildFoldRegionsList(
        element: PsiElement,
        document: Document,
        parent: Expression?,
        overflowGroup: FoldingGroup?,
        overflowLeftPlaceholder: String?,
        overflowRightPlaceholder: String?,
        foldings: MutableList<FoldingDescriptor>
    ) {
        buildFoldRegionsList(element, document, parent, foldings)
    }

    open fun isCollapsedByDefault(): Boolean = true

    open fun isNested(): Boolean = textRange != element.textRange

    open fun isOverflow(): Boolean = isLeftOverflow() || isRightOverflow()

    open fun isLeftOverflow(): Boolean = false

    open fun isRightOverflow(): Boolean = false

    open fun isHighlighted(): Boolean = false

    open fun getHighlightedTextRange(): TextRange = textRange

    override fun toString(): String = element.text

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Expression

        if (element != other.element) return false
        if (textRange != other.textRange) return false

        return true
    }

    override fun hashCode(): Int {
        var result = element.hashCode()
        result = 31 * result + textRange.hashCode()
        return result
    }

    companion object {
        val EMPTY_ARRAY: Array<FoldingDescriptor> = emptyArray()
        const val HIGHLIGHTED_GROUP_POSTFIX: String = ":highlighting"
    }
}
